### ISSUE TITLE
undefined reference to `_ULx86_64_step'

### ISSUE THREAD

**Issue by svenstaro** [PDAL/PDAL#4211]
On Arch Linux, I get this problem compiling PDAL 2.6.0 against libunwind 1.6.2:

```
FAILED: bin/pdal_io_sbet_reader_test
: && /usr/bin/c++ -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/pdal/src=/usr/src/debug/pdal -flto=auto -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto=auto test/unit/CMakeFiles/pdal_test_support.dir/Support.cpp.o test/unit/CMakeFiles/pdal_io_sbet_reader_test.dir/io/SbetReaderTest.cpp.o -o bin/pdal_io_sbet_reader_test  -Wl,-rpath,/build/pdal/src/PDAL-2.6.0-src/build/lib  lib/libpdalcpp.so.16.0.0  lib/libgtest.a && :
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_step'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_Ux86_64_getcontext'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_init_local'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_proc_name'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_reg'
collect2: error: ld returned 1 exit status
```

Compiling PDAL 2.5.6 works just fine. Both done using gcc `gcc (GCC) 13.2.1 20230801`.

**Comment by svenstaro** [comment]
There's this related issue from another project: https://github.com/gnss-sdr/gnss-sdr/issues/639

Might be helpful when debugging this.

**Comment by abellgithub** [comment]
I don't think any developers are using Arch Linux. I know there were some strange things WRT building packages on Arch, but I don't remember the details. You'd have to debug this yourself or at least attach a full build log for anyone to have a clue about what's going on.

**Comment by sixy6e** [comment]
Only just got around to updating PDAL, and yep same issue for me.  Fails in the build process when running tests:

```bash
[457/584] Linking CXX executable bin/pgpointcloudtest
FAILED: bin/pgpointcloudtest
: && /usr/bin/c++ -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now test/unit/CMakeFiles/pdal_test_support.dir/Support.cpp.o plugins/pgpointcloud/CMakeFiles/pgpointcloudtest.dir/test/PgpointcloudTest.cpp.o -o bin/pgpointcloudtest  -Wl,-rpath,/home/me/pkgbuilds/PDAL/pdal/src/PDAL-2.6.0-src/build/lib  lib/libgtest.a  lib/libpdal_plugin_reader_pgpointcloud.so.16.0.0  lib/libpdal_plugin_writer_pgpointcloud.so.16.0.0  /usr/lib/libpq.so  lib/libpdalcpp.so.16.0.0 && :
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_step'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_Ux86_64_getcontext'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_init_local'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_proc_name'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_reg'
collect2: error: ld returned 1 exit status
[458/584] Building CXX object apps/CMakeFiles/pdal.dir/pdal.cpp.o
ninja: build stopped: subcommand failed.
==> ERROR: A failure occurred in build().
    Aborting...
```


And again when building with the TileDB plugin:


```bash
[474/607] Linking CXX executable bin/pdal_io_tiledb_utils_test
FAILED: bin/pdal_io_tiledb_utils_test 
: && /usr/bin/c++ -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now test/unit/CMakeFiles/pdal_test_support.dir/Support.cpp.o plugins/tiledb/CMakeFiles/pdal_io_tiledb_utils_test.dir/test/TileDBUtilsTest.cpp.o plugins/tiledb/CMakeFiles/pdal_io_tiledb_utils_test.dir/io/TileDBUtils.cpp.o -o bin/pdal_io_tiledb_utils_test  -Wl,-rpath,/home/me/pkgbuilds/PDAL/pdal/src/PDAL-2.6.0-src/build/lib:/usr/local/lib  lib/libpdalcpp.so.16.0.0  lib/libgtest.a  /usr/local/lib/libtiledb.so && :
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_step'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_Ux86_64_getcontext'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_init_local'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_proc_name'
/usr/bin/ld: lib/libpdalcpp.so.16.0.0: undefined reference to `_ULx86_64_get_reg'
collect2: error: ld returned 1 exit status
[475/607] Building CXX object apps/CMakeFiles/pdal.dir/pdal.cpp.o
ninja: build stopped: subcommand failed.
==> ERROR: A failure occurred in build().
    Aborting...
```


I can get it to build when setting `-DWITH_BACKTRACE=OFF`.

I've attached the output of the build process if that helps.

[output.log](https://github.com/PDAL/PDAL/files/13445782/output.log)

**Comment by t0b3** [comment]
same linking issue here packaging for gentoo with all the binaries

**Comment by svenstaro** [comment]
For the record, in Arch I now package this without libunwind support. This isn't great but it's not a critical feature and it makes the package build.

**Comment by t0b3** [comment]
FTR: gentoo is heading the same path `-DWITH_BACKTRACE=OFF`

**Comment by jiapei100** [comment]
> For the record, in Arch I now package this without libunwind support. This isn't great but it's not a critical feature and it makes the package build.

@svenstaro 

Does it have something to do with the **libunwind** is from **gcc** or **clang**?

I roughly remember the **libunwind** from/with **gcc** used to conflict against **libunwind** from/with **clang**.

**Comment by jiapei100** [comment]
> -DWITH_BACKTRACE=OFF

yeah... built ... Thank you ...

**Comment by svenstaro** [comment]
> > For the record, in Arch I now package this without libunwind support. This isn't great but it's not a critical feature and it makes the package build.
> 
> @svenstaro
> 
> Does it have something to do with the **libunwind** is from **gcc** or **clang**?
> 
> I roughly remember the **libunwind** from/with **gcc** used to conflict against **libunwind** from/with **clang**.

No idea.

**Comment by hobu** [comment]
> This isn't great but it's not a critical feature and it makes the package build.

The unwind support is mostly a developer tool anyway. Maybe we could consider defaulting it to off, but it would be nice to know why Arch and Gentoo are complainers here. Maybe there is something in how the CMake config detects and links it that's a problem.

**Comment by t0b3** [comment]
[sci-libs:pdal-2.6.0:20231124-130801.log.gz](https://github.com/PDAL/PDAL/files/13464639/sci-libs.pdal-2.6.0.20231124-130801.log.gz)

**Comment by hobu** [comment]
Can you try applying this patch and seeing if it fixes stuff?
```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ae59508e0..abad7dbd7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -260,6 +260,7 @@ target_link_libraries(${PDAL_LIB_NAME}
         ${PDAL_LEPCC_LIB_NAME}
         ${PDAL_LAZPERF_LIB_NAME}
         ${UTFCPP_LIB_NAME}
+        ${LIBUNWIND_LIBRARIES}
     PUBLIC
         ${WINSOCK_LIBRARY}
     INTERFACE
```

**Comment by t0b3** [comment]
> Can you try applying this patch and seeing if it fixes stuff?

unfortunately the situation is still the same with this patch

**Comment by abellgithub** [comment]
I just built PDAL successfully with on Arch LInux with gcc 13.2 - no warnings or errors. `WITH_BACKTRACE` is set to `ON`

> (base) [abell@db8422fa0252 build]$ systemctl --version
systemd 254 (254.6-1-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified

>(base) [abell@db8422fa0252 build]$ uname -a
Linux db8422fa0252 6.4.16-linuxkit #1 SMP PREEMPT_DYNAMIC Thu Nov 16 10:55:59 UTC 2023 x86_64 GNU/Linux

Can you please provide the exact command you're using to build PDAL that's generating these errors?

**Comment by sixy6e** [comment]
Whilst there is a build process that works by including `-DWITH_BACKTRACE=OFF`, I toyed around with various test builds to provide some additional input to the process.

I've used both build options from the AUR:

* [pdal](https://gitlab.archlinux.org/archlinux/packaging/packages/pdal/-/blob/main/PKGBUILD?ref_type=heads)
* [pdal-git](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=pdal-git)

Editing the package scripts to include the TileDB option `-DBUILD_PLUGIN_TILEDB=ON`.

I've also tested by manually running the build commands from the package scripts.

 One uses ninja, the other make:

```bash
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_PLUGIN_PGPOINTCLOUD=ON -DWITH_COMPLETION=ON -DBUILD_PLUGIN_TILEDB=ON
make -j $(nproc --all)
```

And with one processor:

```bash
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_PLUGIN_PGPOINTCLOUD=ON -DWITH_COMPLETION=ON -DBUILD_PLUGIN_TILEDB=ON
make
```

```bash
cmake -Bbuild -GNinja -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_PLUGIN_PGPOINTCLOUD=ON -DWITH_COMPLETION=ON -DBUILD_PLUGIN_TILEDB=ON
ninja -C build
```

I've also tried building using the patch from @hobu, and I still get the same error.

The latest attempt was cloning manually and using the same build steps listed above; with the same outcome.

Anyway, that at least describes what commands were used in the build process.
